api/hypershift/v1beta1/hostedcluster_types.go

```diff
@@ -351,141 +351,161 @@ const (
 	// +kubebuilder:default={networkType: "OVNKubernetes", clusterNetwork: {{cidr: "10.132.0.0/14"}}, serviceNetwork: {{cidr: "172.31.0.0/16"}}}
 	Networking ClusterNetworking `json:"networking"`
 
-	// Autoscaling specifies auto-scaling behavior that applies to all NodePools
-	// associated with the control plane.
+	// autoscaling specifies auto-scaling behavior that applies to all NodePools
+	// associated with this HostedCluster.
 	//
 	// +optional
 	Autoscaling ClusterAutoscaling `json:"autoscaling,omitempty"`
 
-	// Etcd specifies configuration for the control plane etcd cluster. The
-	// default ManagementType is Managed. Once set, the ManagementType cannot be
+	// etcd specifies configuration for the control plane etcd cluster. The
+	// default managementType is Managed. Once set, the managementType cannot be
 	// changed.
 	//
-	// +kubebuilder:validation:Optional
 	// +kubebuilder:default={managementType: "Managed", managed: {storage: {type: "PersistentVolume", persistentVolume: {size: "8Gi"}}}}
 	// +immutable
 	Etcd EtcdSpec `json:"etcd"`
 
-	// Services specifies how individual control plane services are published from
-	// the hosting cluster of the control plane.
-	//
-	// If a given service is not present in this list, it will be exposed publicly
-	// by default.
+	// services specifies how individual control plane services endpoints are published for consumption.
+	// This requires APIServer;OAuthServer;Konnectivity;Ignition.
+	// This field is immutable for all platforms but IBMCloud.
+	// Max is 6 to account for OIDC;OVNSbDb for backward compability though they are no-op.
+	//
+	// +kubebuilder:validation:MaxItems=6
+	// +kubebuilder:validation:MinItems=4
+	// +kubebuilder:validation:ListType=atomic
+	// +kubebuilder:validation:XValidation:rule="self.all(s, !(s.service == 'APIServer' && s.servicePublishingStrategy.type == 'Route') || has(s.servicePublishingStrategy.route.hostname))",message="If serviceType is 'APIServer' and publishing strategy is 'Route', then hostname must be set"
+	// +kubebuilder:validation:XValidation:rule="['APIServer', 'OAuthServer', 'Konnectivity', 'Ignition'].all(requiredType, self.exists(s, s.service == requiredType))",message="Services list must contain at least 'APIServer', 'OAuthServer', 'Konnectivity', and 'Ignition' service types"
+	// +kubebuilder:validation:XValidation:rule="self.filter(s, s.servicePublishingStrategy.type == 'Route' && has(s.servicePublishingStrategy.route) && has(s.servicePublishingStrategy.route.hostname)).all(x, self.filter(y, y.servicePublishingStrategy.type == 'Route' && (has(y.servicePublishingStrategy.route) && has(y.servicePublishingStrategy.route.hostname) && y.servicePublishingStrategy.route.hostname == x.servicePublishingStrategy.route.hostname)).size() <= 1)",message="Each nodePort publishingStrategy 'nodePort' and 'hostname', and each route publishingStrategy 'hostname' must be unique within the Services list."
```

Comment:
From 1.29, you can use optionals, so you could substitute 
```
has(s.servicePublishingStrategy.route) && has(s.servicePublishingStrategy.route.hostname)
```
with
```
s.servicePublishingStrategy.?route.hostname.hasValue()
```
Perhaps doesn't save much in this example but for longer trees that you're traversing looking for values in the future

You can also `orValue("...")` to provide a default if not set